### PR TITLE
[release/1.3] cherry-pick: finalize gradient & mHC composite metrics for internal medicine monitors (#4870)(#4871)

### PR DESCRIPTION
**Pull request by yuweivvv**
Merged in dev: #4870
Merged in dev: #4871
### PR types
Bug fixes

### PR changes
Others

### Description

Cherry-pick #4870 and #4871 to `release/1.3`.

- #4870 `feat: support gradient metric finalization for internal medicine monitors`
  - Adds `InternalMedicineCallback.on_optimizer_begin`, which calls each monitor's `finalize_scaled_grad_metrics(scaler)` so gradient metrics are unscaled/finalized before the optimizer step.
- #4871 `fix: finalize mHC composite metrics on substep end`
  - Adds `InternalMedicineCallback.on_substep_end`, which calls each monitor's `finalize_composite_microbatch()` so composite mHC metrics are settled per microbatch instead of only at step end.

Both hooks early-return when `self._setup_done` is False, and use `getattr(..., None)` so monitors without these methods are unaffected.

Scope: `paddleformers/trainer/trainer_callback.py` only, +18 lines, no behavior change for users who don't enable the internal medicine monitors. Clean cherry-pick, no conflicts.

### CODE REVIEW

File: paddleformers/trainer/trainer_callback.py [hunk 1]
```diff
@@ -1017,6 +1017,24 @@ def on_step_begin(self, args, state, control, **kwargs):
...
+            finalize = getattr(monitor, "finalize_scaled_grad_metrics", None)
+            if finalize is not None:
+                finalize(scaler)
+
+    def on_substep_end(self, args, state, control, **kwargs):
+        if not self._setup_done:
+            return
+
+        for monitor in self._monitor_dict.values():
+            finalize = getattr(monitor, "finalize_composite_microbatch", None)
```
Comment (risemeup1111) [inline paddleformers/trainer/trainer_callback.py:1034]: P1 当前锁定依赖的 PaddleFleet monitor 注册表没有 `mhc_health`，所以 `_monitor_dict` 不会包含实现 `finalize_composite_microbatch` 的 monitor，这个 hook 实际不会执行；梯度累积时的 mHC composite 快照仍可能被后续 micro-batch 覆盖。请同步包含该 monitor/方法的子模块版本，并增加精确依赖下 `gradient_accumulation_steps>1` 的回归测试。

File: paddleformers/trainer/trainer_callback.py [hunk 2]
```diff
@@ -1017,6 +1017,24 @@ def on_step_begin(self, args, state, control, **kwargs):
             if collect is not None:
                 collect()
 
+    def on_optimizer_begin(self, args, state, control, scaler=None, **kwargs):
+        if not self._setup_done:
+            return
+
+        for monitor in self._monitor_dict.values():
+            finalize = getattr(monitor, "finalize_scaled_grad_metrics", None)
```
Comment (risemeup1111) [inline paddleformers/trainer/trainer_callback.py:1025]: P1 当前 release/1.3 锁定的 `llm-internal-medicine` 子模块并未提供 `finalize_scaled_grad_metrics`，因此这里的 capability 检测会静默跳过，`h_res_logits` 梯度指标不会被还原为 unscaled 值。请同步包含该方法的子模块版本（或在本仓库提供实现），并补充精确依赖下启用 AMP monitor 的回归测试，确认回调实际命中。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
| 序号 | 位置 | 优先级 | 状态 |
| --- | --- | --- | --- |
| 1 | [trainer_callback.py:1025](https://github.com/PaddlePaddle/PaddleFormers/pull/4879#discussion_r3773864372) | P1 | 🚧 |
| 2 | [trainer_callback.py:1034](https://github.com/PaddlePaddle/PaddleFormers/pull/4879#discussion_r3773857244) | P1 | 🚧 |

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Review by From00 — APPROVED** [review]
LGTM
